api/v1alpha1/hostedclusterinfrastructure_types.go

```diff
@@ -0,0 +1,357 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type CurrentPhase string
+
+const (
+	PhaseInit   CurrentPhase = "INIT"
+	PhaseInfra  CurrentPhase = "INFRA"
+	PhaseIAM    CurrentPhase = "IAM"
+	PhaseReady  CurrentPhase = "READY"
+	PhaseVerify CurrentPhase = "VERIFY" // or validating??
```

Comment:
why do we need these phases? Conditions already cover much of this (like infra and iam).

Reply:
At the time, at first glance it was to give a better idea of where in the process it was.
init (nothing has run)
INFRA is as described
Then IAM, then READY.

I'm not hard set on keeping it though.

Reply:
The phase information might be useful for UI to handle HCI resources.

Reply:
> At the time, at first glance it was to give a better idea of where in the process it was.

yeah, I don't think there's a right or wrong approach here necessarily. Over the past few years i've seen a trend where new APIs are shifting away from phases to only use conditions. The primary reasoning has been that conditions are more expressive in the info they can output. Over time phases tend to start getting treated like state machines, which can be problematic when new phases are introduced.

> The phase information might be useful for UI to handle HCI resources.

UI and CLI should be able to pick up on conditions. If you want a condition that's just responsible for outputting where in the process a CR is, consider a `Progressing` condition, where Progressing=True has some message that indicates what's going on, like "Provisioning AWS Infra". and Progressing=False means there's nothing left to do.

Reply:
The output looks something like this:
```shell
NAME         TYPE   INFRA           IAM          PROVIDER REF   PHASE
jnp-infra    AWS    AsExpected      AsExpected   AsExpected     READY
jnp-infra2   AWS    AsExpected      AsExpected   AsExpected     READY
jnp-infra3   AWS    NotAsExpected   AsExpected   AsExpected     INFRA
```
In this scenario, the HCI was deleted, but the HostedCluster finalizer is still in place. You have to look at the Infra condition to see the error message that says its stuck waiting on the finalizer to be removed...

The progression for create is INIT > INFRA > IAM > READY.  For destroy its READY > IAM > INFRA > deleted.  Though if the HostedCluster still exists, it displays as above for the time being.

Reply:
Based on the feedback I removed phase.